### DISCUSSION TITLE
NOT CONVERGED for plastic contact

### DISCUSSION THREAD

**Discussion by johanzhuohan** [ContactEngineering/ContactMechanics#64] [Q&A]
#3 
I was trying to estimate how large the pressure needed to make a full contact (fractional contact area = 1) with simple penetration hardness model. However, the calculation does not converge when the pressure, and therefore, the fractional contact area is high. For my case, with the surface of rms gradient = ~0.08 and hardness = 0.1 E*, the calculation does not converge for P > 0.07E*. Is this numerical issue or physically not meaningful to do calculations at such high pressure level?

**Comment by sannant** [comment]
My appologies for the late answer. I need to set up the notifications for the discussions.

As you found out this is related to `https://github.com/ContactEngineering/ContactMechanics/issues/3`.
Polonsky and Keer needs a lot of iterations near full contact. 

In this case it is better to use the CG that uses the gap as variable. 

This is illustrated in https://github.com/ContactEngineering/ContactMechanics/blob/master/examples/constrained_conjugate_gradients.ipynb.

The ideal solver for large contact areas is `ccg_with_restart` on the primal system (primal = gap as variable, while dual = pressures as variable).

The performances are illustrate by @sindhu2993  here:

<img width="738" alt="grafik" src="https://user-images.githubusercontent.com/40853209/164375697-d1d6eb09-6446-484e-bf20-54674080c677.png">
 

Unfortunately we didn't implement ideal plasticity in this version of the conjugate gradient algorithm, and I am not sure from the top of my head whether it is possible at all.

But what we can certainly do is emulate plasticity with a repulsive cohesive law (using the Adhesion package), where I think we would need a bit of strain hardening just for numerical regularization.

I can put together a minimal example doing that if you are interested, but am not sure whether I will find time within the next week.

**Reply by pastewka** [reply]
I'm not sure this would help with penetration hardness near full contact. The problem is that in the penetration hardness model only the region whose presse is not the hardness and not zero is solved by the CG, which can be a small area even at high total contact area. The decisive problem, however, is that it can be zero in which case the solver switches to a simple overrelaxation mechanism which may be the case here.

@johanzhuohan - what does the logger tell you. It will say either "cg" (for conjugate gradient), "sd" (for steepest descent)" or "mix" (for overrelaxation). Is the solver using the "mix" approach?

**Reply by sannant** [reply]
Ah thank you @pastewka  I didn't think about that.

**Reply by johanzhuohan** [reply]
For those high pressure plastic calculations, the first ~5 iterations uses sd, and it switches to mix for the following iterations. The optimization is stuck after certain iterations, and all the variables (e.g., frac. area, total force, and offset) does not change any more.

**Reply by pastewka** [reply]
Can you send me that example (per mail) so I can debug it? That should not happen.

**Reply by johanzhuohan** [reply]
Thank you. Please see the attached file on the email.
